src/indieweb_utils/webmentions/validate.py

```diff
@@ -0,0 +1,159 @@
+            raise WebmentionValidationError("Source is too large.")
+
+    if "text/html" not in request_item.headers["Content-Type"]:
+        raise WebmentionValidationError("This endpoint only supports HTML webmentions.")
+
+    return True
+
+
+def validate_webmention(source: str, target: str, vouch: str = "", vouch_list: List[str] = []) -> bool:        
+    """
+    Check if a webmention is valid.
+
+    :param source: The source URL of the webmention.
+    :type source: str
+    :param target: The target URL of the webmention.
+    :type target: str
+    :param vouch: The vouch URL of the webmention.
+    :type vouch: str
+    :param vouch_list: A list of vouch domains.
+    :type vouch_list: list
+    :return: Boolean to indicate webmention is valid, boolean
+        stating whether the vouch check has passed.
+    :rtype: bool, bool
+    """
+
+    if source.strip("/") == target.strip("/"):
+        raise WebmentionValidationError("Source and target cannot be the same URL.")
+
+    source_protocol = url_parse.urlparse(source).scheme
+    target_protocol = url_parse.urlparse(target).scheme
+
+    if source_protocol not in ["http", "https"]:
+        raise WebmentionValidationError("Source must contain a URL scheme.")
+    
+    if target_protocol not in ["http", "https"]:
+        raise WebmentionValidationError("Target must contain a URL scheme.")
+    
+    # Only allow 3 redirects before raising an error
+    session = requests.Session()
+    session.max_redirects = 3
+
+    validated_headers = False
+
+    try:
+        check_source_size = session.head(source, timeout=5)
+
+        validated_headers = validate_headers(check_source_size)
+    except requests.exceptions.TooManyRedirects:
+        raise WebmentionValidationError("Source redirected too many times.")
+    except requests.exceptions.Timeout:
+        raise WebmentionValidationError("Source timed out.")
+    except:
+        # pass because HEAD request might not be recognised / processed by the client
+        pass
+
+    try:
+        get_source_for_validation = session.get(source).text
+    except Exception as e:
+        raise WebmentionValidationError("Source could not be retrieved.")
+
+    if validated_headers is False:
+        validated_headers = validate_headers(check_source_size)
+
+    if check_source_size.status_code == 410:
+        raise WebmentionIsGone("Webmention source returned 410 Gone code.")
+
+    parse_page = BeautifulSoup(get_source_for_validation, "lxml")
+
+    # get all <link> tags
+    meta_links = parse_page.find_all("link")
+
+    for link in meta_links:
+        # use meta http-equiv status spec to detect 410s https://indieweb.org/meta_http-equiv_status
+        # detecting http-equiv status 410s is required by the webmention spec
+        if link.get("http-equiv", "") == "Status":
+            if link.get("content", "") == "410 Gone":
+                raise WebmentionIsGone("Webmention source returned 410 Gone code.")
```

Comment:
Oh, I like this exception.

Reply:
Thank you! Although maybe a better name would be "WebmentionSourceIsGone"?

Reply:
Yes - that sounds good to me.